Optimize route addition

nightlinus opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This pr optimizes route addition by removing O(n) duplicate check. In case of big monolitic application this starts to bite performance of application initialization. I have moved duplication check to separate class because now it owns hashmap data structure for searching route.

Solves #1

nightlinus committed:
Optimize duplicate routes check

It used to be O(n) to check duplicate, this introduce hashmap structure for route duplicate checks, thus reduce cost of check to O(1)

Signed-off-by: nightlinus <m.a.ogarkov@gmail.com>

geerteltink commented:
Just wondering what router you are using. If you need speed you may want to look into FastRoute and it's caching support: https://docs.mezzio.dev/mezzio/v3/features/router/fast-route/#fastroute-caching-support

nightlinus commented:
> Just wondering what router you are using. If you need speed you may want to look into FastRoute and it's caching support: https://docs.mezzio.dev/mezzio/v3/features/router/fast-route/#fastroute-caching-support

I am using Aura. Thank you for FastRoute caching link, but i should mention that its caching doesn't bypass `RouteCollector` calls and duplication check. I am working on caching routes in my application and there is some problem that as for now routes collection can be taken only from `RouteCollector` and `getRoutes()` is not part of `RouterInterface` so we have 2 places that actually must contain all routes and 2 places to restore from cache, but this is not the case for this pr.

weierophinney commented on src/RouteCollector.php:
```diff
@@ -49,9 +49,15 @@ class RouteCollector
+     */
+    private $duplicateRouteDetector;
+
+    public function __construct(RouterInterface $router, DuplicateRouteDetector $duplicateRouteDetector)
     {
         $this->router = $router;
+        $this->duplicateRouteDetector = $duplicateRouteDetector;
     }
```
```suggestion
    public function __construct(RouterInterface $router, ?DuplicateRouteDetector $duplicateRouteDetector = null)
    {
        $this->router = $router;
        $this->duplicateRouteDetector = $duplicateRouteDetector ?: new DuplicateRouteDectector();
    }
```

Since it has no constructor arguments, making it nullable and providing a default will keep BC for end-users.

weierophinney reviewed (changes requested):
I rebased for you - please pull before making additional changes.

I've noted one change - if you get that in place, I think this will be ready to merge!

weierophinney commented:
Also, @nightlinus — check the travis failures.

nightlinus commented:
done

weierophinney commented on src/RouteCollector.php:
```diff
@@ -49,9 +49,15 @@ class RouteCollector
+     */
+    private $duplicateRouteDetector;
+
+    public function __construct(RouterInterface $router, ?DuplicateRouteDetector $duplicateRouteDetector = null)
     {
         $this->router = $router;
+        $this->duplicateRouteDetector = $duplicateRouteDetector ?? new DuplicateRouteDetector();
```
Since `DuplicateRouteDetector` is marked `final`, and has no dependencies, couldn't we omit the constructor argument for it, and just leave this line?

```suggestion
    public function __construct(RouterInterface $router)
    {
        $this->router = $router;
        $this->duplicateRouteDetector = new DuplicateRouteDetector();
```

weierophinney commented on test/RouteCollectorTest.php:
```diff
@@ -44,7 +48,7 @@ protected function setUp() : void
         $this->router    = $this->prophesize(RouterInterface::class);
         $this->response  = $this->prophesize(ResponseInterface::class);
-        $this->collector = new RouteCollector($this->router->reveal());
+        $this->collector = new RouteCollector($this->router->reveal(), new DuplicateRouteDetector());
```
If we make the change I suggest previously, this can be reverted.

weierophinney commented on test/RouteCollectorTest.php:
```diff
@@ -218,4 +213,81 @@ public function testGetRoutes()
+            $this->collector->get("/bar$item", $this->noopMiddleware);
+        }
+        $baseDuration = microtime(true) - $start;
+        $this->collector = new RouteCollector($this->router->reveal(), new DuplicateRouteDetector());
```
Same remark here.

10n commented on src/RouteCollector.php:
```diff
@@ -49,9 +49,15 @@ class RouteCollector
     public function __construct(RouterInterface $router)
     {
         $this->router = $router;
+        $this->duplicateRouteDetector = new DuplicateRouteDetector();
```
Why don't you move this in  `RouteCollectorFactory`  ?

Also, I suggest you implement an interface, make it invokable  and also create an alias in ConfigProvider.
This would allow any developer to use it as a service and also override it with a local implementation.

nightlinus replied:
This is BC break as was suggested by @weierophinney

pine3ree replied:
I agree, for another reason, about the factory. (I am not sure about the need for an interface for this task, but not against it)
IMO the duplicateRouteDetector instance should also be allowed to be passed in (and to be) null (via factory-(dev|prod)-configuration), thus skipping duplicate checks. Developers should be allowed to control when actually perform it, e.g. in dev-mode only. A default ALWAYS-ON value would not cause any BC break. Kind regards.

Dragomeat replied:
```suggestion
        $this->duplicateRouteDetector = $duplicateRouteDetector ?? new DuplicateRouteDetector();
```

weierophinney replied:
While I agree about the idea of the factory, for BC purposes, it _does_ need to be enabled by default _at the constructor level_, not at the factory level. If a consumer had decided to create their own factory for the route collector previously, they would not know to pass the `DuplicateRouteDetector` instance, and would thus end up in a scenario where duplicate detection is disabled after updating the library within the same major version.

What we _can_ do is use a flag here. Since the logic is marked `final`, we can have a flag that defaults to `true`, indicating the `DuplicateRouteDetector` should be used. Users who want to _disable_ it would pass a boolean `false` for the flag during instantiation of the `RouteCollector`, which would disable that.

We can then also add a config flag for the factory to consume in order to disable it.

This would address the BC concerns, as well as the concerns that developers have a way to disable it for production.

I'm working on that now, @nightlinus ; no need for you to do more work than you already have on this!

nightlinus commented:
Is there something else i should fix before merging @weierophinney ?

samsonasik commented:
@geerteltink could you give reason for closing this PR? or it was an accidentally hit close PR button?

geerteltink commented:
I did not close it. I'm adding automatic releases according the guide. I guess it got automatically closed when removing develop via cli. If I go over the `reopen and comment` button it says develop branch was removed.

We should revise the workflow and delete the develop and master branch as last, after rebasing all open PRs.

/cc @weierophinney

geerteltink commented:
@nightlinus I cannot reopen or rebase your MR. Can you rebase it onto branch `3.2.x` and see if you can reopen this MR?

/edit: Fix the issue and reopened and rebased onto `3.2.x`

Dragomeat commented on src/RouteCollector.php:
```diff
@@ -49,9 +49,15 @@ class RouteCollector
+     */
+    private $duplicateRouteDetector;
+
     public function __construct(RouterInterface $router)
```
```suggestion
    public function __construct(RouterInterface $router, DuplicateRouteDetector $duplicateRouteDetector = null)
```

nightlinus commented:
Fellows, i dont understand what should i do to help this merged or declined. I have done edits as was suggested by @weierophinney, now i get reviews in different directions. Can someone help me to understand what should i do with this PR please? @geerteltink @samsonasik @boesing

weierophinney committed:
feat: allow selectively disabling duplicate route detection

Duplicate route detection is enabled by default, via a flag in the `RouteCollector` class.
If the flag is set, but no duplicate route detector is present in the instance, one is created on the first request to detect duplicate routes.
If disabled, duplicate route detection is not performed.

The `RouteCollector` factory now also checks for a configuration value, `Mezzio\Router\RouteCollector.detect_duplicates`; when present, the value is passed for the optional second constructor argument of the `RouteCollector`.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fixes CS issues flagged by phpcs

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Adds CHANGELOG entry for #3

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
